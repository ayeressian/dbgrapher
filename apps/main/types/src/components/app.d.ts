import "../localization";
import "./import-components";
import { CSSResultGroup, TemplateResult } from "lit";
import { DBGElement } from "./dbg-element";
export default class extends DBGElement {
    #private;
    static get styles(): CSSResultGroup;
    render(): TemplateResult;
    connectedCallback(): void;
    disconnectedCallback(): void;
}
