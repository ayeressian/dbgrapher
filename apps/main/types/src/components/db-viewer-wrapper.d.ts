import { CSSResultGroup, TemplateResult } from "lit";
import { DBGElement } from "./dbg-element";
export default class DbWrapper extends DBGElement {
    #private;
    static get styles(): CSSResultGroup;
    firstUpdated(): void;
    connectedCallback(): void;
    disconnectedCallback(): void;
    render(): TemplateResult;
}
