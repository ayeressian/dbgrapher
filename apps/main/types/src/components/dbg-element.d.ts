import { LitElement } from "lit";
import { AppState } from "../store/reducer";
export declare abstract class DBGElement extends LitElement {
    #private;
    connectedCallback(): void;
    subscribe<StateType>(selector: (state: AppState) => StateType, onUpdate: (subState: StateType, state: AppState) => void, isEqual?: (oldValue: StateType, newValue: StateType) => boolean): void;
    disconnectedCallback(): void;
}
