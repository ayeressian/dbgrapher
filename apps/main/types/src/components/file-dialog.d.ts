import { CSSResultGroup, TemplateResult } from "lit";
import { DBGElement } from "./dbg-element";
export default class extends DBGElement {
    #private;
    firstUpdated(): void;
    static get styles(): CSSResultGroup;
    connectedCallback(): void;
    render(): TemplateResult;
}
