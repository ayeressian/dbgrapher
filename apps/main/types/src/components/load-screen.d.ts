import { CSSResultGroup, TemplateResult } from "lit";
import { DBGElement } from "./dbg-element";
export default class extends DBGElement {
    private view;
    static get styles(): CSSResultGroup;
    connectedCallback(): void;
    render(): TemplateResult;
}
