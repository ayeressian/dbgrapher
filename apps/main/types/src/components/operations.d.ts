import { DbTypeDialogOpenState } from "../store/slices/dialog/db-type-dialog";
import { DbType } from "../db-grapher-schema";
export declare const undo: () => void;
export declare const redo: () => void;
export declare const getDbType: (dbTypeDialogOpenState: DbTypeDialogOpenState) => Promise<DbType>;
export declare const openFile: () => Promise<void>;
export declare const newFile: () => Promise<void>;
export declare const localDrive: () => boolean;
export declare const googleDrive: () => boolean;
export declare const getDialogsAreClosed: () => boolean;
