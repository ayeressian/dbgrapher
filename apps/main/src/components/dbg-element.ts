import { LitElement } from "lit";
import { Unsubscribe } from "redux";
import { AppState } from "../store/reducer";
import { subscribe } from "../subscribe-store";

export abstract class DBGElement extends LitElement {
  #unsubscribes: Unsubscribe[] = [];

  getShadowRoot(): ShadowRoot {
    if (!this.shadowRoot)
      throw new Error(
        "Shadowroot is not available. Most probably Component is not connected yet."
      );
    return this.shadowRoot;
  }

  connectedCallback(): void {
    super.connectedCallback();

    // Update on local change
    this.subscribe(
      (state) => state.localization,
      () => this.updateComplete
    );
  }

  subscribe<StateType>(
    selector: (state: AppState) => StateType,
    onUpdate: (subState: StateType, state: AppState) => void,
    isEqual?: (oldValue: StateType, newValue: StateType) => boolean
  ): void {
    this.#unsubscribes.push(subscribe<StateType>(selector, onUpdate, isEqual));
  }

  disconnectedCallback(): void {
    super.disconnectedCallback();
    this.#unsubscribes.forEach((unsub) => unsub());
  }
}
