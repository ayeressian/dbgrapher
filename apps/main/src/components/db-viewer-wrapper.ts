import { html, css, CSSResultGroup, TemplateResult } from "lit";
import { customElement } from "lit/decorators.js";
import { actions as setSchemaAction, State } from "../store/slices/load-schema";
import { actions as schemaAction } from "../store/slices/schema";
import { actions as tableDialogAction } from "../store/slices/dialog/table-dialog";
import { actions as dbViewerModeAction } from "../store/slices/db-viewer-mode";
import store from "../store/store";
import DbViewerMode from "../store/slices/db-viewer-mode-type";
import { isMac } from "../util";
import DbViewer, {
  TableClickEvent,
  TableDblClickEvent,
  ViewportClickEvent,
  RelationClickEvent,
  Viewport,
  ColumnSchema,
} from "db-viewer";
import { ColumnFkSchema } from "db-viewer";
import { getDriveProvider } from "../drive/factory";
import { AppState } from "../store/reducer";
import { undo, redo, getDialogsAreClosed } from "./operations";
import { DBGElement } from "./dbg-element";
import DbGrapherSchema from "../db-grapher-schema";
import { hintTimed, HintType } from "../store/slices/hint";

@customElement("dbg-db-viewer")
export default class DbWrapper extends DBGElement {
  #resolveLoaded!: (value: PromiseLike<null> | null) => void;
  #loaded: Promise<null> = new Promise(
    (resolve) => (this.#resolveLoaded = resolve)
  );
  #dbViewer!: DbViewer;
  #relationFirstTableName!: string;
  #relationFirstTableRelations!: ColumnSchema[];
  #mode: DbViewerMode = DbViewerMode.None;

  static get styles(): CSSResultGroup {
    return css`
      :host {
        overflow: auto;
      }

      db-viewer {
        height: 100%;
        --font-family: RobotoCondensed;
      }
    `;
  }

  #onTableDblClick = (event: TableDblClickEvent): void => {
    store.dispatch(tableDialogAction.openEdit(event.detail.name));
  };

  #tableCreateListener = (event: ViewportClickEvent): void => {
    store.dispatch(tableDialogAction.openCreate(event.detail));
  };

  #relationFirstClickListener = (event: TableClickEvent): void => {
    this.#dbViewer.removeEventListener(
      "tableClick",
      this.#relationFirstClickListener
    );
    this.#dbViewer.addEventListener(
      "tableClick",
      this.#relationSecondClickListener
    );
    this.#relationFirstTableName = event.detail.name;
    this.#relationFirstTableRelations = this.#getFirstTableRelations();
    if (this.#relationFirstTableRelations.length === 0) {
      store.dispatch(dbViewerModeAction.none());
      store.dispatch(hintTimed(HintType.RelationCreationFailedFK));
    }
  };

  #getFirstTableRelations = () => {
    const schema = this.#dbViewer.getSchema();
    const table = schema!.tables.find(
      (table) => table.name === this.#relationFirstTableName
    );
    return table!.columns.filter(
      (column) => column.pk && (column as ColumnFkSchema).fk == null
    );
  };

  #getTableRelations = (tableName: string) => {
    const schema = this.#dbViewer.getSchema() as DbGrapherSchema;
    const table = schema.tables.find((table) => table.name === tableName);
    return table!.columns.filter(
      (column) => column.pk && (column as ColumnFkSchema).fk == null
    );
  };

  #createRelation = (secondTableName: string): DbGrapherSchema => {
    const schema = this.#dbViewer.getSchema() as DbGrapherSchema;
    const { tables } = schema!;
    // const secondTable = tables.find((table) => table.name === secondTableName);
    const firstTable = tables.find(
      (table) => table.name === this.#relationFirstTableName
    );

    const secondTableRelations = this.#getTableRelations(secondTableName);

    secondTableRelations.forEach((column) => {
      const originalRelationName = `fk_${secondTableName}_${column.name}`;
      let relationName = originalRelationName;
      let index = 0;
      while (
        firstTable?.columns.find((column) => column.name === relationName) !=
        null
      ) {
        ++index;
        relationName = `${originalRelationName}_${index}`;
      }
      firstTable?.columns.push({
        name: relationName,
        uq:
          this.#mode === DbViewerMode.RelationOneToOne ||
          this.#mode === DbViewerMode.RelationZeroToOne,
        nn:
          this.#mode === DbViewerMode.RelationOneToOne ||
          this.#mode === DbViewerMode.RelationOneToMany,
        fk: {
          table: secondTableName,
          column: column.name,
        },
      });
    });
    return schema!;
  };

  #relationSecondClickListener = (event: TableClickEvent): void => {
    const secondTableName = event.detail.name;
    const schema = this.#createRelation(secondTableName);
    store.dispatch(schemaAction.set(schema));
    store.dispatch(setSchemaAction.loadViewportUnchange());
    store.dispatch(dbViewerModeAction.none());
    this.#dbViewer.removeEventListener(
      "tableClick",
      this.#relationSecondClickListener
    );

    const secondTableRelations = this.#getTableRelations(secondTableName);
    if (secondTableRelations.length === 0) {
      store.dispatch(dbViewerModeAction.none());
      store.dispatch(hintTimed(HintType.RelationCreationFailedFK));
    }
  };

  #onTableMoveEnd = (): void => {
    store.dispatch(
      schemaAction.set(this.#dbViewer.getSchema() as DbGrapherSchema)
    );
    void getDriveProvider().updateFile();
  };

  firstUpdated(): void {
    this.#dbViewer = this.shadowRoot!.querySelector<DbViewer>("db-viewer")!;
    this.#dbViewer.addEventListener("tableDblClick", this.#onTableDblClick);
    this.#dbViewer.addEventListener("tableMoveEnd", this.#onTableMoveEnd);
    this.#resolveLoaded(null);
  }

  #loadSchema = (state: AppState, viewport: Viewport): void => {
    void this.#loaded.then(() => {
      this.#dbViewer.setSchema(state.schema.present, viewport);
      store.dispatch(setSchemaAction.loaded());
    });
  };

  #loadSchemaCheck = (): void => {
    const state = store.getState();
    switch (state.loadSchema) {
      case State.LOAD:
        this.#loadSchema(state, "centerByTables");
        break;
      case State.LOAD_VIEWPORT_UNCHANGE:
        this.#loadSchema(state, "noChange");
        break;
    }
  };

  #onKeyDown = (event: KeyboardEvent): void => {
    if (getDialogsAreClosed() && event.key === "z") {
      if (
        !event.shiftKey &&
        ((event.ctrlKey && !isMac) || (event.metaKey && isMac))
      ) {
        undo();
      }

      if (
        (event.shiftKey &&
          ((event.ctrlKey && !isMac) || (event.metaKey && isMac))) ||
        (!isMac && event.ctrlKey)
      ) {
        redo();
      }
    }
  };

  connectedCallback(): void {
    super.connectedCallback();

    this.#loadSchemaCheck();

    document.addEventListener("keydown", this.#onKeyDown);

    this.subscribe((state) => state.loadSchema, this.#loadSchemaCheck);

    this.subscribe(
      (state) => state.dbViewerMode,
      (dbViewerMode) => {
        this.#mode = dbViewerMode;
        switch (dbViewerMode) {
          case DbViewerMode.CreateTable:
            this.#dbViewer.addEventListener(
              "viewportClick",
              this.#tableCreateListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationFirstClickListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationSecondClickListener
            );
            this.#dbViewer.removeEventListener("tableClick", this.#removeTable);
            this.#dbViewer.removeEventListener(
              "relationClick",
              this.#removeRelation
            );
            break;
          case DbViewerMode.RelationOneToOne:
          case DbViewerMode.RelationZeroToOne:
          case DbViewerMode.RelationZeroToMany:
          case DbViewerMode.RelationOneToMany:
            this.#dbViewer.addEventListener(
              "tableClick",
              this.#relationFirstClickListener
            );
            this.#dbViewer.removeEventListener(
              "viewportClick",
              this.#tableCreateListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationSecondClickListener
            );
            this.#dbViewer.removeEventListener("tableClick", this.#removeTable);
            this.#dbViewer.removeEventListener(
              "relationClick",
              this.#removeRelation
            );
            break;
          case DbViewerMode.Remove:
            this.#dbViewer.addEventListener("tableClick", this.#removeTable);
            this.#dbViewer.addEventListener(
              "relationClick",
              this.#removeRelation
            );
            this.#dbViewer.removeEventListener(
              "viewportClick",
              this.#tableCreateListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationFirstClickListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationSecondClickListener
            );
            break;
          default:
            this.#dbViewer.removeEventListener(
              "viewportClick",
              this.#tableCreateListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationFirstClickListener
            );
            this.#dbViewer.removeEventListener(
              "tableClick",
              this.#relationSecondClickListener
            );
            this.#dbViewer.removeEventListener("tableClick", this.#removeTable);
            this.#dbViewer.removeEventListener(
              "relationClick",
              this.#removeRelation
            );
            break;
        }
      }
    );
  }

  disconnectedCallback(): void {
    super.disconnectedCallback();

    this.#dbViewer.removeEventListener(
      "viewportClick",
      this.#tableCreateListener
    );
    this.#dbViewer.removeEventListener(
      "tableClick",
      this.#relationFirstClickListener
    );
    this.#dbViewer.removeEventListener(
      "tableClick",
      this.#relationSecondClickListener
    );
    this.#dbViewer.removeEventListener("tableClick", this.#removeTable);
    this.#dbViewer.removeEventListener("relationClick", this.#removeRelation);

    document.removeEventListener("keydown", this.#onKeyDown);
  }

  #removeTable = (event: TableClickEvent): void => {
    const schema = this.#dbViewer.getSchema() as DbGrapherSchema;
    const tableName = event.detail.name;
    schema.tables = schema.tables.filter((table) => table.name !== tableName);
    schema.tables.forEach((table) => {
      table.columns = table.columns.filter(
        (column) => (column as ColumnFkSchema).fk?.table !== tableName
      );
    });
    store.dispatch(schemaAction.set(schema));
    getDriveProvider().updateFile();
    store.dispatch(setSchemaAction.loadViewportUnchange());
  };

  #removeRelation = (event: RelationClickEvent): void => {
    const schema = this.#dbViewer.getSchema() as DbGrapherSchema;
    const fromTable = schema.tables.find(
      (table) => table.name === event.detail.fromTable
    )!;
    fromTable.columns = fromTable.columns.filter(
      (column) => column.name !== event.detail.fromColumn
    );
    store.dispatch(schemaAction.set(schema));
    void getDriveProvider().updateFile();
    store.dispatch(setSchemaAction.loadViewportUnchange());
  };

  render(): TemplateResult {
    return html`<db-viewer></db-viewer>`;
  }
}
