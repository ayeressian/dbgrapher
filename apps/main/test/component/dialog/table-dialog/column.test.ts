import "../../../../src/components/import-components";
import Columns from "../../../../src/components/dialog/table-dialog/columns";
import { initComponentTest, removeElement } from "../../../helper";
import { describe, it, beforeEach, afterEach } from "vitest";
import toMatchHtmlSnapshot from "../../../to-match-html-snapshot";

describe("table-dialog-column", () => {
  let tableDialogColumns: Columns;
  let shadowRoot: ShadowRoot;

  beforeEach(async (): Promise<void> => {
    tableDialogColumns = await initComponentTest({
      elementType: "dbg-table-dialog-columns",
      noUpdate: true,
      attrs: {
        tableIndex: "0",
      },
    });
    tableDialogColumns.schema = {
      tables: [
        {
          name: "test",
          columns: [],
          pos: {
            x: 0,
            y: 0,
          },
        },
      ],
    };
    await tableDialogColumns.updateComplete;
    shadowRoot = tableDialogColumns.getShadowRoot();
  });
  afterEach((): void => {
    removeElement(tableDialogColumns);
  });

  it("should render properly", () => {
    toMatchHtmlSnapshot(shadowRoot?.innerHTML);
  });
});
