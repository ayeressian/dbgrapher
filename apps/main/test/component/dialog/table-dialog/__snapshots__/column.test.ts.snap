// Vitest Snapshot v1

exports[`table-dialog-column > should render properly 1`] = `
" <div class=\\"container\\">
<form class=\\"pure-form\\">
<div class=\\"title\\">Columns</div>
<datalist id=\\"types\\">
<option value=\\"CHARACTER\\"></option><option value=\\"VARCHAR\\"></option><option value=\\"NCHAR\\"></option><option value=\\"NCHAR VARYING\\"></option><option value=\\"BINARY\\"></option><option value=\\"BINARY VARYING\\"></option><option value=\\"BINARY LARGE OBJECT\\"></option><option value=\\"NUMERIC\\"></option><option value=\\"DECIMAL\\"></option><option value=\\"SMALLINT\\"></option><option value=\\"INTEGER\\"></option>
<option value=\\"BIGINT\\"></option><option value=\\"FLOAT\\"></option><option value=\\"REAL\\"></option><option value=\\"DOUBLE PRECISION\\"></option><option value=\\"BOOLEAN\\"></option><option value=\\"DATE\\"></option><option value=\\"TIME\\"></option><option value=\\"TIMESTAMP\\"></option><option value=\\"INTERVAL\\"></option></datalist>
<div class=\\"table-container\\">
<table class=\\"pure-table pure-table-horizontal\\">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th title=\\"Primary key\\">PK</th>
<th title=\\"Unique\\">UQ</th>
<th title=\\"Not null\\">NN</th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>

<tr>
<td class=\\"no-column\\" colspan=\\"8\\">No columns to show</td>
</tr></tbody>
</table>
</div>
<button class=\\"pure-button add-column\\">
Add column
</button>
</form>
</div><style>
table {
width: 760px;
}
undefined
undefined
undefined
.add-column {
margin-top: 10px;
}
td.no-column {
text-align: center;
color: #AAA;
}
.container {
margin-top: 30px;
}
.title {
margin-bottom: 10px;
}
.table-container {
max-height: 200px;
overflow-x: hidden;
overflow-y: auto;
display: inline-block;
}
.table-container table th {
background-color: #E0E0E0;
position: sticky;
top: 0;
}
.table-container table {
border-collapse: collapse;
}
.add-column {
display: block;
}
table tbody td:last-child {
width: 32px;
}
.remove-icon {
width: 32px;
height: 32px;
background-image: url(/asset/remove.svg);
background-size: cover;
}
.move-up-icon {
width: 32px;
height: 32px;
background-image: url(/asset/move-up.svg);
background-size: cover;
}
.move-up-icon.disabled {
filter: opacity(30%);
}
.move-down-icon {
width: 32px;
height: 32px;
background-image: url(/asset/move-down.svg);
background-size: cover;
}
.move-down-icon.disabled {
filter: opacity(30%);
}
}
</style>"
`;
